storage/innobase/log/log0recv.cc

```diff
@@ -2698,43 +2698,57 @@ recv_sys_t::parse_mtr_result recv_sys_t::parse(source &l, bool if_exists)
       mach_write_to_4(iv + 8, space_id);
       mach_write_to_4(iv + 12, page_no);
     }
+    DBUG_PRINT("ib_log",
+               ("scan " LSN_PF ": rec %x len %zu page %u:%u",
+                lsn, b, l - recs + rlen, space_id, page_no));
     got_page_op= !(b & 0x80);
-    if (!got_page_op);
-    else if (storing == BACKUP && srv_operation == SRV_OPERATION_BACKUP)
-    {
-      if (page_no == 0 && (b & 0xf0) == INIT_PAGE && first_page_init)
-        first_page_init(space_id);
-      continue;
-    }
-    else if (storing == YES && file_checkpoint &&
-             space_id != TRX_SYS_SPACE && !srv_is_undo_tablespace(space_id))
+    if (got_page_op)
     {
-      recv_spaces_t::iterator i= recv_spaces.lower_bound(space_id);
-      if (i != recv_spaces.end() && i->first == space_id);
-      else if (lsn < file_checkpoint)
-        /* We have not seen all records between the checkpoint and
-        FILE_CHECKPOINT. There should be a FILE_DELETE for this
-        tablespace later. */
-        recv_spaces.emplace_hint(i, space_id, file_name_t("", false));
-      else
+      if (storing == BACKUP)
       {
-        const page_id_t id(space_id, page_no);
-        if (!srv_force_recovery)
+        if (page_no == 0 && (b & 0xf0) == INIT_PAGE && first_page_init)
+          first_page_init(space_id);
+        else if (rlen == 1 && undo_space_trunc)
         {
-          ib::error() << "Missing FILE_DELETE or FILE_MODIFY for " << id
-                      << " at " << lsn
-                      << "; set innodb_force_recovery=1 to ignore the record.";
-          goto corrupted;
+          mach_write_to_4(iv + 8, space_id);
+          mach_write_to_4(iv + 12, page_no);
+          byte buf;
+          if (*l.copy_if_needed(iv, &buf, recs, 1) == TRIM_PAGES)
+            undo_space_trunc(space_id);
```

Comment:
On Windows, we got this:
```
Error:Run-Time Check Failure #2 - Stack around the variable 'buf' was corrupted. At C:\Buildbot\amd64-windows\build\storage\innobase\log\log0recv.cc:3137
```
I now realize that `copy_if_needed()` requires a larger buffer than just the payload. For other modes than `storage=BACKUP` we allocate `decrypt_buf` from the stack. I would want to limit the stack size for `mariadb-backup --backup` here.. The worst case could around 20 bytes here. I’ll try to determine the exact value and revise this.